### PR TITLE
A binary RBM example

### PR DESCRIPTION
**Pull request by asitstands**
## Description ##

An example of binary restricted Boltzmann machine learning MNIST.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by yifeim** [comment]
Very cool implementation! Helped me refresh on the RBM literature. Both test scripts make sense to me.

A few comments:
* It would be awesome to include some basic mathematical descriptions in the README. I got my intuitions from http://deeplearning.net/tutorial/rbm.html . However, you have a clearer backward function that is not explicit in the tutorial (even though I believe in your intuitions.)

* An alternative idea on the sampler function. I would not compute prob_0 after getting prob_1; instead, to sample from Bernoulli distributions:
`mx.nd.random.uniform(shape=shape_of_prob_1) < prob_1`

* Would it be possible to track the training process by reporting CD-loss after each epoch? In a similar vein, an additional feature to consider is to include aux in the forward loop for persistent-CD.

* Minor styling ideas: Naming `interaction_weight` instead of `interaction` can receive the extra benefit of default initialization. Also, since the spirit of the commit is a CustomOperator, it may be simpler to use gluon SymbolBlock instead of HybridBlock?

**Comment by yifeim** [comment]
To simplify the comments, I would love to see
*documentation, e.g., the link I provided in the previous comment. 
*sampler rewrite for 2x speed up. 
*training loss tracked at the end of each epoch. 

The other ideas are only suggestions and may be out of scope in this request.

**Comment by asitstands** [comment]
@yifeim Thanks for the suggestions!

- The gradient in this example is the standard one from the straightforward calculation of the gradient of the log likelihood, to be used in the stochastic gradient descent (SGD). The expressions in the tutorial you linked replace the average over the Gibbs samples of hidden units with the probability of a hidden unit to be one. They are interchangeable in practice, while strictly speaking the latter is not exact for SGD.
- There are lots of literature on RBM and each of them has its own presentation of the math. No one may directly match with the code in this PR. Github does not support math rendering with the markdown so explaining the math of RBM in the README is difficult and I think that is also out of scope of this PR. If someone wants to understand the example code (not merely using), then they should understand the math and I think that the differences in the literature may not be hard to be adjusted.
- Yes, the Bernoulli sampling method you provide is the standard idiom! I forgot to restore it after some tests with `multinomial` ( #10970). Thanks.
- `interaction_weight` sounds good. I'll use the name.
- Monitoring the progress of the RBM learning is not trivial because the calculation of the loss (objective function) is practically impossible. There is no golden standard for the monitoring as far as I know. So I left it empty. But anyway adding one to this example would not be bad. My personal preference is using the annealed importance sampling and some people recommend it (e.g. [a guide by Hinton](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwiwwPHs5PLbAhWKi7wKHWIiCOwQFggrMAA&url=https%3A%2F%2Fwww.cs.toronto.edu%2F~hinton%2Fabsps%2FguideTR.pdf&usg=AOvVaw2--ZzyoAVNqnBr993kZeF8)). I'll add it.

I'll update the code soon.

**Comment by yifeim** [comment]
Thanks for the explanations.

I did not mean to expect a full paper in README.md -- that would indeed be redundant; instead I  just wanted some reference links to your favorite papers. Also a good idea is to highlight the results. For example, your current demo samples new images from raw images -- I was wondering if both images could be included in README.

I see that RBM is the case where computing the gradient is easier than computing the actual likelihood. For AIS, the algo box and (15)-(17) in http://homepages.inf.ed.ac.uk/imurray2/pub/08dbn_ais/dbn_ais.pdf may be a good starting point. If you decide to implement AIS, please highlight the expected AIS values as well.

**Comment by szha** [comment]
@asitstands thanks for updating the PR. @yifeim would you mind taking another pass at this PR?

**Comment by asitstands** [comment]
Now the log-likelihoods of the test and training data are reported at the completion of each epoch. They are estimated using AIS. README shows some samples generated from the RBM.

**Comment by yifeim** [comment]
Hi @asitstands ,

Thanks for the update! Sorry I did not notice the code changes until was reminded. If you don't mind, I would like to do a full run through on Monday. I feel this is an awesome contribution in general and it will set high standards for other people to understand custom operators and RBMs.

Thanks,
Yifei

**Comment by asitstands** [comment]
@yifeim It was my fault that the update was late. I forgot to push the update and went on a vacation :grin:. I pushed the ais implementation and some tweaks yesterday.

I found that the gluon implementation is much slower than the module implementation. nvidia-smi reports about 60% of GPU utilization for the module implementation but only 35% for the gluon implementation. If I replace gluon's `DataLoader` with `NDArrayIter`, the gluon implementation's performance becomes comparable to the module implementation. I'm not sure that this performace issue is unavoidable with `DataLoader` or I have made some mistakes. Would you please check this performance issue?

**Comment by asitstands** [comment]
I set `num_workers=4` for the `DataLoader` used in the gluon implementation. This improves the performance but it is still slower than the module implementation (about 35% longer running time).

**Comment by yifeim** [comment]
Just an update that I ran the symbol codes a couple days back and the presentation seemed much clearer than the original submission. The simulated images also improved in the sense that more sampling steps were taken.

I had a little issue that the gluon codes did not run on my windows machine. Could be a platform related issue and would love to check on a linux platform.

**Comment by yifeim** [comment]
rbm_example
The comparison at the end of 130 epochs. I think it looks very nice!

File: example/restricted-boltzmann-machine/binary_rbm_gluon.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            param_visible_layer_bias, param_hidden_layer_bias, param_interaction_weight,
+            args.ais_batch_size, args.ais_num_batch, args.ais_intermediate_steps, args.ais_burn_in_steps, train_data, ctx)
+    print("Epoch %d completed with test log-likelihood %f and train log-likelihood %f" % (epoch, test_log_likelihood, train_log_likelihood))
+
+
+### Show some samples.
+
+# Each sample is obtained by 3000 steps of Gibbs sampling starting from a real sample.
+# Starting from the real data is just for convenience of implmentation.
+# There must be no correlation between the initial states and the resulting samples.
```
[the remark this answers is not in the document]
Comment (asitstands) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: Why irrelevant? I chose starting from real images insted of random images, but starting from a random images is also a common practice. Mentioning an alternative choice would be helpful here. No correlation between the initial state and the the steady state is also a general property of a Markov chain.

[the remark this answers is not in the document]
Comment (yifeim) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: Precisely, starting from real images is a common practice, whereas starting from random states is not common. 

Let's see if the following arguments make sense: The negative examples in the training loop all start from real images. The training loop may never see some true negatives from uniform random choices, because the number of visited states is far fewer than the 2^130 possibilities. As a result, swapping the sampling seed with a uniform random start may have unforeseeable outcomes.

As a matter of fact, I did the experiment with random pixel values sampled from an iid beta distribution. More than half of the images did not converge to images of digits.

In the paper I referenced, the authors trained an VAE using real images. However, visualization of the latent states reveals large areas that are not visited by the real images. Sampling from them would likely generate unreal digits. I believe this is a common issue with generative models.

[the remark this answers is not in the document]
Comment (asitstands) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: I'm not sure that the result in the paper you mentioned has how much generality for various generative models (Anyway the paper itself looks interesting). I think the trained RBM hardly define a reducible Gibbs chain. If it does, training an RBM using Gibbs sampling could be very difficult problem. As we increase the running time of the Gibbs sampling in the training, the quality of the resulting model gets better. This is a common observation, but I think it is also unexpected if the chain is reducible. Thus the Gibbs chain may be irreducible or has very limited number of isolated states. Then the initial state and the steady state of the chain must have no direct relation. If they have some relation, that means the running time of the chain is too short. 

The relaxation to the steady state is exponential except some special cases. To reach the steady state, the chain does not visit each possible image in the uniform probability. Also it does not need to visit a large fraction of the all possible images. The relaxation is exponential but the constant factor could matter in reality. So we start the chain from the real images. Starting from real images is just a trick to reduce the long relaxation time of the random initial state. Anyway it is a matter of the constant factor. It cannot beat the exponential nature of the relaxation. If you get images far from digits in the samples obtained from a *long* Gibbs chain starting from random images, they must finally appear in the samples starting from the real images. So the problem is just how long we can run chain.

From the same argument, I think comparing the initial real images and the resulting images is also not appropriate. It could be misleading. If you want to see the quality of the trained RBM, comparing some random samples from the test dataset and the resulting images of the model would be an appropriate way.

By the way, there are many papers with good RBM samples starting from random states. For example, the AIS paper you linked also shows examples of the RBM samples starting from randomly initialized MNIST images.

[the remark this answers is not in the document]
Comment (asitstands) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: I think that the probability distribution used to generate the random initial state could also matter. I'm not sure that you casted the random numbers from the beta distribution to binary units. If not, I guess it could make the relaxation time longer because much larger fraction of the state space has very low probability and starting from such states needs longer time to reach the steady state. Basically this RBM assumes the state space of binary units.

[the remark this answers is not in the document]
Comment (yifeim) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: Ok. I am convinced. The speed of convergence does not depend on the total number of states, but rather how fast any state transitions to any other state. It would be nice if there are analyses applied to RBMs in particular.

Recommend to approve unaltered.

[the remark this answers is not in the document]
Comment (szha) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: @yifeim are the previous batch of reviews still relevant?

[the remark this answers is not in the document]
Comment (yifeim) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: To answer your question, I was using grayscale on the image pixel values. I guess you were taking float numbers to initialize but then use binary values on the samples?

[the remark this answers is not in the document]
Comment (yifeim) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: @szha The comments are not necessary at this point. We could publish the original submission to save everyone's time.

[the remark this answers is not in the document]
Comment (yifeim) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: @asitstands Actually, what do you think about including the png that compares the original and the sampled - so that people can understand the differences and that the posteriors truly mixed?

[the remark this answers is not in the document]
Comment (asitstands) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: @yifeim Yes I mean the initialization with grayscale or black/white. The real data is grayscale, but anyway the model assumes binary pixels. This discrepancy can be ignored in many case but sometimes it makes difference. Anyway I think it is not importance in this stage.

I didn't think about demonstrating the mixing. It would be meaningful. Ok, I'll update the code to include the starting images as you suggested. That could show the quality of the trained model and also show the mixing.

[the remark this answers is not in the document]
Comment (yifeim) [reply example/restricted-boltzmann-machine/binary_rbm_gluon.py]: Sounds good to me and thanks for the efforts!

**Comment by asitstands** [comment]
I updated the code. Now it shows the initial real images also. The sample image included in the README is the one provided by yifeim. Thanks @yifeim.
